### ISSUE TITLE
Reinforcement Learning

### ISSUE THREAD

**Issue by christopher-wilke** [huggingface/candle#1065]
Does `candle` also offers support for Reinforcement Learning?

**Comment by LaurentMazare** [comment]
There is nothing really specific for this in candle but it's the same for most ML framework that I would know of. E.g. in `tch` we didn't need any specific bits in order to build [these examples](https://github.com/LaurentMazare/tch-rs/tree/main/examples/reinforcement-learning). It might be possible to build some dedicated RL library on top of candle but that probably wouldn't belong to the core set of crates - that said incorportating some simple RL examples would probably be a good idea.

**Comment by LaurentMazare** [comment]
Tagging with "help-wanted". If anyone wanted to build some reinforcement learning based examples for candle, this would be great.
For example one could start by porting the `a2c` and `ppo` examples from tch-rs (see [here](https://github.com/LaurentMazare/tch-rs/tree/main/examples/reinforcement-learning)) and use it on the same atari games. The tch-rs and candle api are very similar so this should hopefully not be too difficult.

**Comment by dashdeckers** [comment]
It might take me a bit as I'm fairly new to Rust, but I'd like to take a stab at implementing / porting over the DDPG algorithm and setting up some tests / demos via your gymnasium wrapper PR that I just noticed :)

**Comment by LaurentMazare** [comment]
Ah actually I started with the gymnasium wrapper but plan on porting the RL bits from `tch-rs` ([see here](https://github.com/LaurentMazare/tch-rs/tree/main/examples/reinforcement-learning)). So DDPG is the first one that I was planning to port to check that everything works out properly. Maybe there are other RL algorithms that you may want to take a stab at?

**Comment by dashdeckers** [comment]
Fair enough :) I'm actually working part-time on a (masters) thesis involving DDPG so I'll be sticking to that one, but I can wait for your push since there are other to-dos on my list and then contribute the improvements / additions I end up implementing along the way.

**Comment by LaurentMazare** [comment]
Ah, great that you're doing your thesis on DDPG. I would say that if you think it's going to be a week or two, then you can go ahead and make a DDPG example, if it's likely to be more I would rather have a working example soon so that other folks can start playing with RL - and you can obviously also make your own implementation on the side.

**Comment by dashdeckers** [comment]
It's likely going to take longer than a week or two. I have one too many projects going on right now and trying not to compromise on the work-life balance, so I'd rather not make such guarantees :) 

I salute your pace and progress with Candle btw!

**Comment by dashdeckers** [comment]
I've tried my hand at restructuring the ReplayBuffer and I'd appreciate some feedback, is this a good place to post & discuss code snippets or do you prefer another channel?

**Comment by LaurentMazare** [comment]
Great, here works well for me.

**Comment by dashdeckers** [comment]
I thought a VecDeque would be an appropriate data structure, and instead of keeping track of several of them I grouped them into a Transition struct. 

I left a comment in the hopes that you may know a nice way to "unzip" an iterator into several buckets in a single pass instead of iterating again for each bucket. 

Also I am unsure whether my design introduces any other performance inefficiencies (e.g. Tensor::cat). 

```rust
use std::collections::VecDeque;

use candle::{DType, Device, Tensor};
use rand::{distributions::Uniform, thread_rng, Rng};

#[derive(Clone)]
struct Transition {
    obs: Tensor,
    action: Tensor,
    reward: Tensor,
    next_obs: Tensor,
    terminated: bool,
    truncated: bool,
}
impl Transition {
    fn new(
        obs: &Tensor,
        action: &Tensor,
        reward: &Tensor,
        next_obs: &Tensor,
        terminated: bool,
        truncated: bool,
    ) -> Self {
        Self {
            obs: obs.clone(),
            action: action.clone(),
            reward: reward.clone(),
            next_obs: next_obs.clone(),
            terminated,
            truncated,
        }
    }
}

struct ReplayBuffer {
    buffer: VecDeque<Transition>,
    capacity: usize,
    size: usize,
}

impl ReplayBuffer {
    fn new(capacity: usize) -> Self {
        Self {
            buffer: VecDeque::with_capacity(capacity),
            capacity,
            size: 0,
        }
    }

    fn push(
        &mut self,
        obs: &Tensor,
        action: &Tensor,
        reward: &Tensor,
        next_obs: &Tensor,
        terminated: bool,
        truncated: bool,
    ) {
        if self.size == self.capacity {
            self.buffer.pop_front();
        } else {
            self.size += 1;
        }
        self.buffer.push_back(Transition::new(
            obs, action, reward, next_obs, terminated, truncated,
        ));
    }

    fn random_batch(&self, batch_size: usize) -> Option<(Tensor, Tensor, Tensor, Tensor)> {
        if self.size < batch_size {
            None
        } else {
            let transitions: Vec<Transition> = thread_rng()
                .sample_iter(Uniform::from(0..self.size))
                .take(batch_size)
                .map(|i| self.buffer.get(i).unwrap().clone())
                .collect();

            let obs: Vec<Tensor> = transitions.iter().map(|t| t.obs.clone()).collect();
            let action: Vec<Tensor> = transitions.iter().map(|t| t.obs.clone()).collect();
            let reward: Vec<Tensor> = transitions.iter().map(|t| t.obs.clone()).collect();
            let next_obs: Vec<Tensor> = transitions.iter().map(|t| t.obs.clone()).collect();

            //
            // would be nice to be able to unzip the iterator into the four buckets at once?
            // (without importing something like unzip_n)
            // then we could replace the above with the following:
            //

            // let (obs, action, reward, next_obs) = thread_rng()
            //     .sample_iter(Uniform::from(0..self.size))
            //     .take(batch_size)
            //     .map(|i| self.buffer.get(i).unwrap().clone())
            //     .map(|t| (t.obs, t.action, t.reward, t.next_obs))
            //     .collect::<(Vec<Tensor>, Vec<Tensor>, Vec<Tensor>, Vec<Tensor>)>();

            Some((
                Tensor::cat(&obs, 0).unwrap(),
                Tensor::cat(&action, 0).unwrap(),
                Tensor::cat(&reward, 0).unwrap(),
                Tensor::cat(&next_obs, 0).unwrap(),
            ))
        }
    }
}
```

If this looks fine to you, then I will continue on to the actual DDPG algorithm.

**Comment by LaurentMazare** [comment]
I think it's a great start and `VecDeque` makes a lot of sense here. I don't have a better suggestion re unzipping but your code looks fine, there might be some small typos around the `t.obs` though that you want to adapt.
This indeed will not be super efficient using `Tensor::cat` but to start with it's probably good and once you get something working we can profile it to see where the time is actually being spent and optimize these bits.

**Comment by dashdeckers** [comment]
Oh yeah, good catch! Alright, thanks for the feedback.

**Comment by dashdeckers** [comment]
Can you point me to some resources explaining / motivating the `VarMap` / `VarBuilder` APIs?

In `tch-rs` there is the `VarStore` which (in the DDPG example) is saved by the model structs (I guess to facilitate cloning) and the optimizer as well as the layers (via some steps through `Path`) are registered there automatically.

In the Mnist example in `candle`, the `VarBuilder` is only passed to the constructor functions as an argument and things are registered there (somehow?), and the `VarMap` is not passed down to the models at all. Also the `VarBuilder` knows the Device and DType so that doesn't need to be saved by the model struct either I would guess?

I guess I'm trying to understand and wrap my head around the concepts, which are new to me, to figure out how to handle them in the most idiomatic way and I could use some more reading material or examples.

For instance, in my first draft for the Actor constructor, I am unsure whether I should be passing down both the `VarBuilder` and the `VarMap` (since I need the `VarMap` to initialize the optimizer and the `VarBuilder` to initialize the layers) or should I construct the `VarBuilder` within the constructor on the fly?

```rust
struct Actor {
    network: Sequential,
    optimizer: Optimizer,
    num_obs: usize,
    num_actions: usize,
    learning_rate: f64,
}
impl Actor {
    fn new(
        varbuilder: VarBuilder,
        varmap: VarMap,
        num_obs: usize,
        num_actions: usize,
        learning_rate: f64,
    ) -> Self {

        // let varmap = VarMap::new();
        // let vb = VarBuilder::from_varmap(&varmap, DType::F32, &Device::Cpu);

        let optimizer = AdamW::new(
            varmap.all_vars(),
            ParamsAdamW {
                lr: learning_rate,
                ..Default::default(),
            },
        ).unwrap();

        let prefix = "actor";
        let network = seq()
            .add(linear(num_obs, 400, varbuilder.pp(concat!(prefix, "fc1"))))
            .add(Activation::Relu)
            .add(linear(400, 300, varbuilder.pp(concat!(prefix, "fc2"))))
            .add(Activation::Relu)
            .add(linear(300, num_actions, varbuilder.pp(concat!(prefix, "fc3"))))
            .add(Activation::Sigmoid); // TODO: This should be Tanh.

        Self {
            network,
            optimizer,
            num_obs,
            num_actions,
            learning_rate,
        }
    }
}
```

**Comment by LaurentMazare** [comment]
Sorry there isn't much documentation here.
The `VarMap` actually stores the variables, it can be used to serialize and deserialize them.
A `VarBuilder` can be queried for some variables (and maintains some notion of path). It's a very thin wrapper over a backend that may either load the variables from disk, take them from memory, initialize them, or do whatever. In the `VarMap` case new variables get initialized.
When building a model, you want to pass the `VarBuilder` around. This abstracts what is used to retrieve the variables so the same code can be used either to create a model for training or for inference with pre-trained weights. The `VarMap` is only typically used at top level, as you mentioned to pass all the variables to the optimizer, save the variables in a checkpoint (or load them from a checkpoint).

**Comment by dashdeckers** [comment]
I'm trying to create a copy of a network, so for example with a Linear layer I am calling `.get` on a `VarBuilder`, to get the weights Tensor, then I `.copy()` that Tensor and create a new `Linear` layer explicitly with `::new()`. However I still need to register this Tensor on the `VarMap` right? But I don't see a way to do that, so I'm thinking I'm probably going about this the wrong way...

I want to be able to make a copy of a `Sequential` network and give everything a specific prefix because I use the prefix to register only that part of the network on an `Optimizer` like so:

```rust
fn filter_varmap_by_prefix(varmap: VarMap, prefix: &str) -> Vec<Var> {
    varmap.data().lock()
        .unwrap()
        .iter()
        .filter_map(|(name, var)| name.starts_with(prefix).then_some(var.clone()))
        .collect::<Vec<_>>()
}

let actor_optim = AdamW::new(
    filter_varmap_by_prefix(varmap.clone(), "actor-target"),
    ParamsAdamW {
        lr: actor_lr,
        ..Default::default()
    },
)?;
```

Along the way I'm also trying to translate this code over:

```rust
fn track(dest: &mut nn::VarStore, src: &nn::VarStore, tau: f64) {
    tch::no_grad(|| {
        for (dest, src) in
            dest.trainable_variables().iter_mut().zip(src.trainable_variables().iter())
        {
            dest.copy_(&(tau * src + (1.0 - tau) * &*dest));
        }
    })
}
```

So far I have something like this, but I need to insert a `Var` and not a `Tensor` into varmap..

```rust
let layers: Vec<Linear> = Vec::new();
for (i, (in_dim, out_dim)) in zip(1..=3, [(self.num_obs, 400), (400, 300), (300, self.num_actions)]) {
    let weight = vb.get((out_dim, in_dim), &format!("{source_prefix}-fc{i}.weight"))?.copy()?;
    let bias = vb.get(in_dim, &format!("{source_prefix}-fc{i}.bias"))?.copy()?;

    varmap.get((in_dim, out_dim), &format!("{target_prefix}-fc{i}.weight"), Default::default(), vb.dtype(), vb.device());

    varmap.data().lock().unwrap().insert(format!("{target_prefix}-fc{i}.weight"), weight).unwrap();
    varmap.data().lock().unwrap().insert(format!("{target_prefix}-fc{i}.bias"), bias).unwrap();

    let layer = Linear::new(weight, Some(bias));
    layers.push(layer)
}

```

**Comment by LaurentMazare** [comment]
Ah these very reinforcement learning specific workflows are certainly not well supported on candle at the moment, great time to polish this and make it easier :)
What I would suggest there is relying on the `set` method on variables which should be the equivalent of PyTorch `copy_` (but only is available on `Var` and not on `Tensor`).
Do you think you would be able to implement `track` with this? Ideally you would have two varmap only and don't need to add new variables on each step but rather just copy the contents.

Your `filter_varmap_by_prefix` sounds good. Note that you can probably avoid a bit of cloning by using references and also rely on `all_vars` rather than `data()` would be simpler to extract the variables and then filter them.

Let me know if I can be of any more help (or if I misunderstood some of the difficulties you mentioned :) ).

**Comment by dashdeckers** [comment]
Sorry for the pause, had to focus on work again for a few days. I'll be focusing again on this for the next few days. 

I have a (very ugly) version of `track` going and it seems to work, and it seems like the weights are updating but (of course) it's not improving / learning correctly. Is there a way to check which variables are registered on an optimizer to sanity check that the rights things are being updated?

**Comment by dashdeckers** [comment]
Also I just watched the bias of the last layer of the critic (which is a single float) update and change over time and then suddenly go from `~ -0.035..` to `NaN`, do you have an idea what could cause that?

**Comment by LaurentMazare** [comment]
> Is there a way to check which variables are registered on an optimizer to sanity check that the rights things are being updated?

You can print the variables when passing them to the optimizers. That won't guarantee that they get updated though if the loss doesn't depend on them (or if there is a bug in candle). Maybe you can use a model that is simple enough so that you can just print the tensor values and check that they change?

**Comment by dashdeckers** [comment]
It works!! I'll make a pull request in a bit

**Comment by dashdeckers** [comment]
Oh I'm not sure what is going on with my pull request to be honest?

**Comment by LaurentMazare** [comment]
Yeah you may want to transplant your changes to the current git tip and make a clean one instead as it seems pretty odd.

**Comment by dashdeckers** [comment]
Still seems odd right?
Then I will make a clean one...

**Comment by dashdeckers** [comment]
Okay sorry about that! Now it should be fine

**Comment by LaurentMazare** [comment]
Closing this now that we have some RL example.
